net/udprelay/server.go

```diff
@@ -0,0 +1,517 @@
+	} else {
+		pair[0] = discoA
+		pair[1] = discoB
+	}
+	return pair
+}
+
+// ServerEndpoint contains the Server's endpoint details.
+type ServerEndpoint struct {
+	// ServerDisco is the Server's Disco public key used as part of the 3-way
+	// bind handshake. Server will use the same ServerDisco for its lifetime.
+	// ServerDisco value in combination with LamportID value represents a
+	// unique ServerEndpoint allocation.
+	ServerDisco key.DiscoPublic
+
+	// LamportID is unique and monotonically increasing across ServerEndpoint
+	// allocations. It enables clients to dedup and resolve allocation event
+	// order. Clients may race to allocate on the same Server, and signal
+	// ServerEndpoint details via alternative channels, e.g. DERP. Additionally,
+	// Server.AllocateEndpoint() requests may not result in a new allocation
+	// depending on existing server-side endpoint state. Therefore, where
+	// clients have local, existing state that contains ServerDisco and
+	// LamportID values matching a newly learned endpoint, these can be
+	// considered one and the same. If ServerDisco is equal, but LamportID is
+	// unequal, LamportID comparison determines which ServerEndpoint was
+	// allocated most recently.
+	LamportID uint64
+
+	// AddrPorts are the IP:Port candidate pairs the Server may be reachable
+	// over.
+	AddrPorts []netip.AddrPort
+
+	// VNI (Virtual Network Identifier) is the Geneve header VNI the Server
+	// will use for transmitted packets, and expects for received packets
+	// associated with this endpoint.
+	VNI uint32
+
+	// BindLifetime is amount of time post-allocation the Server will consider
+	// the endpoint active while it has yet to be bound via 3-way bind handshake
+	// from both client parties.
+	BindLifetime time.Duration
+
+	// SteadyStateLifetime is the amount of time post 3-way bind handshake from
+	// both client parties the Server will consider the endpoint active lacking
+	// bidirectional data flow.
+	SteadyStateLifetime time.Duration
+}
+
+// serverEndpoint contains Server-internal ServerEndpoint state. serverEndpoint
+// methods are not thread-safe.
+type serverEndpoint struct {
+	// discoPubKeys contains the key.DiscoPublic of the served clients. The
+	// indexing of this array aligns with the following fields, e.g.
+	// discoSharedSecrets[0] is the shared secret to use when sealing
+	// Disco protocol messages for transmission towards discoPubKeys[0].
+	discoPubKeys       pairOfDiscoPubKeys
+	discoSharedSecrets [2]key.DiscoShared
+	handshakeState     [2]disco.BindUDPRelayHandshakeState
+	addrPorts          [2]netip.AddrPort
+	lastSeen           [2]time.Time
+	challenge          [2][disco.BindUDPRelayEndpointChallengeLen]byte
+
+	lamportID   uint64
+	vni         uint32
+	allocatedAt time.Time
+}
+
+func (e *serverEndpoint) handleDiscoControlMsg(from netip.AddrPort, senderIndex int, discoMsg disco.Message, uw udpWriter, serverDisco key.DiscoPublic) {
+	handshakeState := e.handshakeState[senderIndex]
```

Comment:
Check for panic necessary? Same for a few other places in this function?

Reply:
`senderIndex` is always 0 or 1, but I went ahead and added a guard on `senderIndex` values.